### PR TITLE
improving error handling for explore tab graphs (SCP-3160)

### PR DESCRIPTION
**Pull request by devonbush**
Ooof, this was tougher than expected...  I created a standardized way of indicating errors in visualizations (see screenshot below), and extended it to our four main visualization types.  This also does some overriding of morpheus error handling so we can show a better error message when the user does a gene search with no results (previously a dialog with 'Error opening https:...' would show, which implied a connection problem, rather than an empty search)

This also ran `yarn lint --fix` to fix some outstanding code format issues

To test:

1. Load a study, with the react_explore feature flag on

2. Do a single gene search for a gene that doesn’t exist (e.g. ‘xxxxxx’).  confirm error message is display, similar to screenshot below

3. Do a multi-gene search where all genes don’t exist (e.g. ‘ddddddd’, ‘dfsadfasdf’).  confirm error message is displayed similar to below.

image

### CODE REVIEW

File: app/controllers/api/v1/visualization/clusters_controller.rb
Comment (bistline) [inline app/controllers/api/v1/visualization/clusters_controller.rb:108]: (puts on HTTP status code hat) `422` is `Unprocessable Entity`, and deals specifically the case when the request had the correct syntax/instructions, but the server could not process them.  I've really only ever seen it applied to a `POST` or `PUT` when saving data, and the parameters supplied cause a validation error, in which case the server responds `422` because the request could not be completed as outlined.

In this particular case, we're trying to find some visualization data via a `GET`, but `ClustersController.get_cluster_viz_data` doesn't find what we asked for.  I would posit then that the correct response code would be `404` or `400`, though it would depend on what the request parameters actually were.  If we were missing a required parameter, `400` would be more correct, but if the sum of all params equates to an empty response, then `404`.

File: app/controllers/api/v1/visualization/expression_controller.rb
Comment (bistline) [inline app/controllers/api/v1/visualization/expression_controller.rb:123]: This to me a is `400` - we're missing a required parameter.

File: app/controllers/api/v1/visualization/expression_controller.rb
Comment (bistline) [inline app/controllers/api/v1/visualization/expression_controller.rb:125]: Again, I'd opt for `404` here.

File: app/javascript/components/explore/ExploreDisplayTabs.js
Comment (bistline) [inline app/javascript/components/explore/ExploreDisplayTabs.js:454]: Probably not in the context of this PR, but I really think we should refactor this method for readability/maintainability.  In context with #922 we now have 8 different tabs that can be rendered at any given time, and the cyclomatic complexity of this is such I had no idea what the logic was going to be here without a serious 10 minutes of inspection and copying snippets out to a text document to assemble what it "would" be after the merge.  This clearly isn't going to scale well, so I think it's time for a refactor.

File: app/javascript/components/visualization/Heatmap.js
Comment (bistline) [inline app/javascript/components/visualization/Heatmap.js:120]: Interesting - I'd love a dive on `withErrorBoundary` at some point to better understand how it works.

File: app/javascript/lib/error-message.js
Comment (bistline) [inline app/javascript/lib/error-message.js:48]: I would say that perhaps we should talk to Josh about this - there may be something we're not doing in terms of setup that can assist with this.

**Review by bistline — CHANGES_REQUESTED** [review]
Some quibbles about HTTP status codes, and a question regarding `getEnabledTabs`.  Otherwise looks good - I can tell this wasn't simple.

**Review by bistline — APPROVED** [review]

File: app/javascript/components/visualization/DotPlot.js
Comment (eweitz) [inline app/javascript/components/visualization/DotPlot.js:83]: Blocker: this naming pattern will confuse maintainers.  

The default export is `SafeDotPlot`, but it's imported as `DotPlot` -- the same name as the "unsafe" function in this module.  To avoid confusion over similar names, please align the imported and exported names -- e.g. name the unwrapped dot plot function "UnsafeDotPlot", then:

```suggestion
const DotPlot = withErrorBoundary(UnsafeDotPlot)
export default DotPlot
```

And similarly for the other visualization components.

This would also make it obvious when using the unsafe version of the function.

File: app/javascript/components/visualization/StudyViolinPlot.js
Comment (eweitz) [inline app/javascript/components/visualization/StudyViolinPlot.js:65]: For readability, unpack this condition as in `ScatterPlot.js`.

File: app/javascript/components/visualization/ScatterPlot.js
Comment (eweitz) [inline app/javascript/components/visualization/ScatterPlot.js:48]: For readability, unpack this condition:

```suggestion
   const apiOK = checkScpApiResponse(clusterResponse, () => Plotly.purge(graphElementId), setShowError, setErrorContent)
    if (apiOK) {
```

File: app/javascript/lib/error-message.js
Comment (eweitz) [inline app/javascript/lib/error-message.js:8]: Less runny, and more conventional form:

```suggestion
/** 
  * Hook for enabling the conditional display an error message.
  * In the event of an error, the returned ErrorComponent can be placed in the component's markup, 
  * and will be displayed with the content from setErrorContent if setShowError is called with true.
  */
```

**Review by eweitz — CHANGES_REQUESTED** [review]
I'm a fan of moving towards more inline error UI, as done here!  They're sometimes trickier (as seen in demo), but overall better UI.

Two blockers:

* See comment beginning "Blocker: this naming pattern will confuse maintainers."

* This degrades error observability relative to PR #885.  On staging, do a gene search for "asdfasdf".  Note how an `error-modal` event is logged to Bard / Mixpanel.  No analogous client-side error logging is done here.  Even if the event names differ a bit, we should preserve similar functionality.

**Review by eweitz — APPROVED** [review]

**Pull request merged**
